### ISSUE TITLE
`NullNode` injected when instantiating data class with nullable property

### ISSUE THREAD

**Issue by varahash** [FasterXML/jackson-databind#2705]
Following tests fails:

```
class Test {

    data class MyData(val value: JsonNode?)

    @Test
    fun testNullValue() {
        val mapper = ObjectMapper()
        mapper.registerModule(KotlinModule())

        val json = "{\"value\": null}"

        val actual: MyData = mapper.readValue(json, MyData::class.java)

        val expected = MyData(null)
        Assert.assertEquals(expected, actual)
    }

    @Test
    fun testMissingProperty() {
        val mapper = ObjectMapper()
        mapper.registerModule(KotlinModule())

        val json = "{}"

        val actual: MyData = mapper.readValue(json, MyData::class.java)

        val expected = MyData(null)
        Assert.assertEquals(expected, actual)
    }
}
```

**Comment by viartemev** [comment]
@cowtowncoder from my point of view it makes sense to serialize absent of value to the proper presentation of the type but, **it should be documented somewhere**.
For example, [`Optional` deserialized to `Optional.empty` in case of value absent](https://github.com/FasterXML/jackson-module-kotlin/issues/204), what makes sense.

**Comment by cowtowncoder** [comment]
Agreed. Documentation is a big challenge. But there may also be semantic component, as `JsonNode` is not (or I don't think of it as) quite same "reference type" as `Optional`.
I can see this going either way.

On documentation, I suppose this is in area of "null handling", in a way.
But more generally there probably should be some sort of table for various Features for different Java types, their general handling. My biggest problem is usually that of getting started, creating structure for documentation: it is easier to fill in structure with more information, incrementally, than come up with a good structure to start.

Another related challenge is that much of this is controlled in decentralized manner, basically by `JsonSerializer`/`JsonDeserializer` implementing certain handlers.

**Comment by cowtowncoder** [comment]
Actually fixed for #3214 (in Jackson 2.13.0) -- missing (absent) value for `JsonNode` is basic Java/Kotlin `null`.

There may be improvements to be had in future (maybe map `MissingNode`, optionally), but for now this is solved.
